### ISSUE TITLE
spring_layout iterations converge to bad layouts, not spread out

### ISSUE THREAD

**Issue by joseortiz3** [networkx/networkx#3325]
Here is my reproducible example. Nevermind where the data came from:

```python
import networkx as nx
import numpy as np
import matplotlib.pyplot as plt

edges = ((16103, 4423, {'sel_frac': 0.10104208846993425}), (58975, 4423, {'sel_frac': 0.11392450913277707}), (43510, 41220, {'sel_frac': 0.10052701693748936}), (45700, 41220, {'sel_frac': 0.11816510197282526}), (79350, 44800, {'sel_frac': 0.09355756316139689}), (15345, 50090, {'sel_frac': 0.10798202394419384}), (25610, 50090, {'sel_frac': 0.13080150443192698}), (21700, 51630, {'sel_frac': 0.13016626070338505}), (74510, 51630, {'sel_frac': 0.14163814697449248}), (75660, 86300, {'sel_frac': 0.10210561210316822}))

G = nx.DiGraph()
G.add_edges_from(edges)

# Plotting
pos = nx.spring_layout(G,iterations=5000)

def edge_attr(key):
    return np.array(tuple(e[2][key] for e in G.edges(data=True)))

edge_widths = edge_attr('sel_frac')
edge_widths = edge_widths / max(edge_widths)

edge_labels = dict([((u,v,),('%0.2f'%d['sel_frac']))
                 for u,v,d in G.edges(data=True)])

pos2 = {key: value + np.array([0,0.06]) for key, value in pos.items()}

nx.draw_networkx_nodes(G, pos)
nx.draw_networkx_labels(G, pos2)
nx.draw_networkx_edges(G,pos,width=edge_widths,labels=edge_labels)
nx.draw_networkx_edge_labels(G,pos,edge_labels=edge_labels)
plt.show()
```

![figure_1](https://user-images.githubusercontent.com/18585219/53324070-1d0e2d00-389d-11e9-9b5f-bf9195c789e5.png)

As you can see, this directed network's layout converges undesirably, because all the connected nodes stack on top of each-other. This must mean the "physical forces" specified in the implementation aren't doing their job.

**Comment by dschult** [comment]
Yes -- this is a result of multiple components and too many iterations.
If you reduce the number of iterations to 50, it looks much better.

Basically what is happening is that the components are drifting apart as the iterations increase.
When the distance apart swamps the distance between node in the component, then all nodes in the component appear to be at the same position.

I think the short term fix is to reduce the number of iterations. The longer term fix involves changing the spring layout to somehow provide spring-like action between components.  Might be best to find a published algorithm that handles this before diving into it.

**Comment by joseortiz3** [comment]
I don't know any published algorithms handling this case, but I suppose I can find the most-connected node in each component ("disjoint subgraph") and fix those nodes in a radial layout, with the radius proportional to `k`, the equilibrium distance in the `spring layout`.

**Comment by joseortiz3** [comment]
I added a force term to the code, and spring pulling every node to the origin, with force `pos / (k * np.sqrt(nnodes)`. I figured since `k` is the equilibrium distance, then `k**2` is the equilibrium area, and therefore the entire plot having area `nnodes * k**2` implies a length scale `np.sqrt(nnodes) * k`.

```python
        # displacement "force"
        displacement = np.einsum('ijk,ij->ik',
                                 delta,
                                 (k * k / distance**2 - A * distance / k))
        # prevent things from flying off into infinity if not connected
        displacement = displacement - pos / ( k * np.sqrt(nnodes))
```

This results in convergence after 5000 iterations to
![image](https://user-images.githubusercontent.com/18585219/53377558-aa926100-391f-11e9-885b-5edf8087acb2.png)

**Comment by dschult** [comment]
Yes -- it's probably better to draw the components separately. You might take a look at
[nxviz](https://github.com/ericmjl/nxviz) to see if it has better treatment for multiple components. It is a visualization package for NetworkX.

**Comment by joseortiz3** [comment]
I think the "center spring", attaching a spring between every node and the center of the plot, is a sensible, extremely useful, and simple-to-implement feature that I think should be in `spring_layout`'s optional arguments/parameters; `def spring_layout(G, center_spring=False)`, implemented exactly as I described earlier. I'll put together a pull request for your consideration.

**Comment by rossbar** [comment]
This one's been inactive for awhile and I think is sufficiently addressed (at least for the typical use-cases) with reducing the number of iterations as suggested.

For reference, here's what the original example looks like with the default number of iterations (50):

![draw_spring_components](https://user-images.githubusercontent.com/1268991/176260729-0b9c3dc6-fcb6-4974-80c7-1e3b3a2e69d6.png)

Force-directed layout algorithms that are specifically tailored for graphs with multiple components are of course always welcome!
